src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,156 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
             }
         }
 
+        /// <inheritdoc />
+        public override async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (TryReadInternal(tokenSource, out ReadResult readResult))
```

Comment:
This code still isn't as optimized as it should be. I outlined what I expected the implementation should look like in the comments

Reply:
I have tried another optimization step, tell me whether it is in the right direction...
I am not pleased with the use of `AdvanceTo(segment, segment.End, segment, segment.End)`, do you see a better alternative?

Reply:
I think you'd only pay for it once per call to CopyToAsync right?